modules/javafx.controls/src/test/java/test/javafx/scene/control/ControlContractTest.java

```diff
@@ -0,0 +1,127 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package test.javafx.scene.control;
+
+import javafx.beans.property.DoubleProperty;
+import javafx.beans.property.SimpleDoubleProperty;
+import javafx.geometry.Insets;
+import javafx.scene.Scene;
+import javafx.scene.control.SkinBase;
+import javafx.stage.Stage;
+import org.junit.jupiter.api.AfterEach;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.ValueSource;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class ControlContractTest {
+
+    private ControlStub control;
+
+    @BeforeEach
+    void setup() {
+        control = new ControlStub();
+    }
+
+    @AfterEach
+    void cleanup() {
+        control = null;
+    }
+
+    /**
+     * Tests the contract of {@link SkinBase#layoutChildren(double, double, double, double)} method that is called
+     * when a {@link javafx.scene.control.Control} should layout itself and his children.
+     * The x, y, width and height should be snapped according to the render scale.
+     * The width and height should subtract the snapped padding insets.
+     *
+     * @param scale the render scale
+     */
+    @ParameterizedTest
+    @ValueSource(doubles = { 1.0, 1.25, 1.5, 1.75, 2.0 })
+    void testLayoutChildrenContract(double scale) {
+        DoubleProperty renderScaleProperty = new SimpleDoubleProperty(scale);
+
+        Stage stage = new Stage();
+        stage.renderScaleXProperty().bind(renderScaleProperty);
+        stage.renderScaleYProperty().bind(renderScaleProperty);
+
+        control.setMinSize(200, 300);
+        control.setPadding(new Insets(4, 5, 6, 7));
+
+        AtomicBoolean layoutChildrenCalled = new AtomicBoolean(false);
+
+        control.setSkin(new SkinBase<>(control) {
+
+            @Override
+            protected void layoutChildren(double x, double y, double w, double h) {
+                layoutChildrenCalled.set(true);
+
+                Insets padding = control.getPadding();
+
+                double expectedX = control.snappedLeftInset();
+                assertEquals(expectedX, x);
+
+                expectedX = control.snapPositionX(padding.getLeft());
+                assertEquals(expectedX, x);
+
+                double expectedY = control.snappedTopInset();
+                assertEquals(expectedY, y);
+
+                expectedY = control.snapPositionY(padding.getTop());
+                assertEquals(expectedY, y);
+
+                double expectedWidth = control.snapSizeX(control.getWidth()) - control.snappedLeftInset() - control.snappedRightInset();
+                assertEquals(expectedWidth, w);
```

Comment:
since we are performing floating point operations, we might accumulate small error.  
So this and 3 subsequent assertEquals() would need an EPSILON = 0.0000001 (an arbitrary value)

Reply:
Will have a look, but it worked without so I did not bothered adding an epsilon to the tests

Reply:
I would have added an EPSILON.  I suspect we just got lucky that with the default padding and a small subset of scales everything comes out ok.